### PR TITLE
Centered Knowledge sprite

### LINKED ISSUE #4835: Adventure map, Arena window, modify the Knowledge sprite so that the book is more centered in the frame
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

In fheroes2, the book of the Knowledge sprite is not centered horizontally in the frame.

In fheroes2:

Arena level sprite 01

### Describe the possible solution

Modify the Knowledge sprite so that the book is shifted horizontally by one pixel to the left, so that it is more centered in the frame.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by felix642**
closes #4835 

Here's what it looks like with my changes:
image

### CODE REVIEW

**Comment by LeHerosInconnu** [comment]
Hello @felix642,

Nice. :)
Would it be possible to copy the first column of pixels on the left of the original graphic and paste it on the last column of pixels on the right of the modified graphic in this change to avoid having this "doubling" of the last two columns of pixels on the right of the graphic of this change.

In fheroes2:

Open the three images in different tabs in your browser, enlarge the images and switch between the three tabs alternately.
First, original Knowledge, second, Knowledge with this modification, third, Knowledge with the first column of pixels copied and pasted:
Arena original knowledge 01 Arena modified knowledge 01 Arena first pixel column used knowledge 01

**Comment by felix642** [comment]
Hi @LeHerosInconnu,
I added the changes that you requested. It's hard to tell if I did perfectly since it is just a line of pixel and I would appreciate if you could take a look at it to make sure that everything looks perfect now.

**Comment by LeHerosInconnu** [comment]
Hello @felix642,

> Hi @LeHerosInconnu, I added the change that you requested. It's hard to tell if I did perfectly since it is just a line of pixel and I would appreciate if you could take a look at it to make sure that everything looks perfect now.

Almost. :)

The graphic inside the frame is 35 columns wide.
The pixel column was pasted on the second to last column (column 34) to the left of the yellow frame of the Knowledge graphic instead of the last column (column 35) just to the left of the yellow frame. :)

In fheroes2:

Open the two images in different tabs in your browser, enlarge the images and switch between the two tabs alternately.
First, Knowledge with current modification, second, Knowledge with the first column of pixels copied and pasted:
Arena first pixel column used second knowledge 01 Arena first pixel column used knowledge 01

**Comment by LeHerosInconnu** [comment]
Hello @felix642,

I checked, everything is fine.
Thanks for the change. :)

File: src/fheroes2/dialog/dialog_arena.cpp [hunk 1]
```diff
@@ -162,8 +162,17 @@ void InfoSkillClear( const fheroes2::Rect & rect1, const fheroes2::Rect & rect2,
     fheroes2::Blit( fheroes2::AGG::GetICN( ICN::XPRIMARY, 1 ), display, rect2.x, rect2.y );
     fheroes2::Blit( fheroes2::AGG::GetICN( ICN::XPRIMARY, 2 ), display, rect3.x, rect3.y );
 
-    if ( Settings::Get().ExtHeroArenaCanChoiseAnySkills() )
-        fheroes2::Blit( fheroes2::AGG::GetICN( ICN::XPRIMARY, 3 ), display, rect4.x, rect4.y );
+    if ( Settings::Get().ExtHeroArenaCanChoiseAnySkills() ) {
+        constexpr int32_t BORDER_WIDTH = 2;
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_arena.cpp:166]: Please use `const` over `constexpr` and name the variable in camelCase.

Comment (ihhub) [reply src/fheroes2/dialog/dialog_arena.cpp:166]: The same applies for the line 194.

File: src/fheroes2/dialog/dialog_arena.cpp [hunk 2]
```diff
@@ -162,8 +162,17 @@ void InfoSkillClear( const fheroes2::Rect & rect1, const fheroes2::Rect & rect2,
...
+        const fheroes2::Sprite & knowledgeICN = fheroes2::AGG::GetICN( ICN::XPRIMARY, 3 );
+        const fheroes2::Sprite & croppedICN
+            = fheroes2::Crop( knowledgeICN, BORDER_WIDTH + 1, BORDER_WIDTH, rect4.width - 2 * BORDER_WIDTH - 1, rect4.height - 2 * BORDER_WIDTH );
+        const fheroes2::Sprite & pixelLine = fheroes2::Crop( knowledgeICN, BORDER_WIDTH, 0, 1, rect4.height );
+
+        fheroes2::Blit( knowledgeICN, display, rect4.x, rect4.y );
+        fheroes2::Blit( croppedICN, display, rect4.x + BORDER_WIDTH, rect4.y + BORDER_WIDTH );
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_arena.cpp:173]: Please use `knowledgeICN` directly without cropping an image. `fheroes2::Blit` function has other variant just to cover this case.

File: src/fheroes2/dialog/dialog_arena.cpp [hunk 3]
```diff
@@ -162,8 +162,17 @@ void InfoSkillClear( const fheroes2::Rect & rect1, const fheroes2::Rect & rect2,
...
+        fheroes2::Blit( pixelLine, display, rect4.x + rect4.width - BORDER_WIDTH - 1, rect4.y );
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_arena.cpp:174]: Please use `knowledgeICN` directly without cropping an image. `fheroes2::Blit` function has other variant just to cover this case.

File: src/fheroes2/dialog/dialog_arena.cpp [hunk 4]
```diff
@@ -181,8 +190,18 @@ void InfoSkillSelect( int skill, const fheroes2::Rect & rect1, const fheroes2::R
...
+        if ( Settings::Get().ExtHeroArenaCanChoiseAnySkills() ) {
+            constexpr int32_t BORDER_WIDTH = 2;
+            const fheroes2::Sprite & knowledgeICN = fheroes2::AGG::GetICN( ICN::XPRIMARY, 7 );
+            const fheroes2::Sprite & croppedICN
+                = fheroes2::Crop( knowledgeICN, BORDER_WIDTH + 1, BORDER_WIDTH, rect4.width - 2 * BORDER_WIDTH - 1, rect4.height - 2 * BORDER_WIDTH );
+            const fheroes2::Sprite & pixelLine = fheroes2::Crop( knowledgeICN, BORDER_WIDTH, 0, 1, rect4.height );
+
+            fheroes2::Blit( knowledgeICN, display, rect4.x, rect4.y );
+            fheroes2::Blit( croppedICN, display, rect4.x + BORDER_WIDTH, rect4.y + BORDER_WIDTH );
+            fheroes2::Blit( pixelLine, display, rect4.x + rect4.width - BORDER_WIDTH - 1, rect4.y );
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_arena.cpp:202]: Please use `knowledgeICN` directly without cropping an image. `fheroes2::Blit` function has other variant just to cover this case.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @felix642 , I left multiple comments here. Could you please take a look?

File: src/fheroes2/dialog/dialog_arena.cpp (hunk 1 above)
Comment (felix642) [reply src/fheroes2/dialog/dialog_arena.cpp:166]: I'm sorry, I'm just a bit curious. What is the motivation behind using const over constexpr in this kind of situation ?

Comment (ihhub) [reply src/fheroes2/dialog/dialog_arena.cpp:166]: `constexpr` is only useful when a compiler needs to calculate something at the time of compilation. Here we have nothing to calculate.

File: src/fheroes2/dialog/dialog_arena.cpp [hunk 5]
```diff
@@ -162,8 +162,15 @@ void InfoSkillClear( const fheroes2::Rect & rect1, const fheroes2::Rect & rect2,
     fheroes2::Blit( fheroes2::AGG::GetICN( ICN::XPRIMARY, 1 ), display, rect2.x, rect2.y );
     fheroes2::Blit( fheroes2::AGG::GetICN( ICN::XPRIMARY, 2 ), display, rect3.x, rect3.y );
 
-    if ( Settings::Get().ExtHeroArenaCanChoiseAnySkills() )
-        fheroes2::Blit( fheroes2::AGG::GetICN( ICN::XPRIMARY, 3 ), display, rect4.x, rect4.y );
+    if ( Settings::Get().ExtHeroArenaCanChoiseAnySkills() ) {
+        const int32_t BORDER_WIDTH = 2;
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_arena.cpp:166]: Hi @felix642 , please use camelCase variable naming here and for the same variable below.

**Review by ihhub — CHANGES_REQUESTED** [review]

**Comment by ihhub** [comment]
@felix642 , huge thanks for fixing this issue!

**Pull request merged**